### ISSUE TITLE
Programmatically add V8 breakpoint to top of main file on --debug-brk

### ISSUE THREAD

**Issue by CrabDude** [babel/babel#2565]
**TLDR; `babel-node` should ensure the initial breakpoint is the first line of application code when `--debug-brk` is set.**

In pursuit of `babel-node` being a drop-in replacement for `node` in all (non-production) regards to further encourage adoption...

A breakpoint should be added programmatically to the top of the main application file when `--debug-brk` is set. Debugging `babel-node` with `node-inspector` by default break(point)s on `lib/_babel-node.js`, which is inconvenient. Libraries like [`babel-node-debug`](https://babeljs.io/docs/setup/#babel_node_debug) (I am the author) alleviate this a bit by programmatically setting a breakpoint at the top of the main file. (Some discussion and details [here](https://github.com/CrabDude/babel-node-debug/issues/6))

However, it's currently **not possible** to accomplish the following workflow without resolving this issue:

1. `$ babel-node --debug-brk=8010 foo.js`
1. `$ node-inspector`
1. Navigate to `http://127.0.0.1:8080/?ws=127.0.0.1:8080&port=8010`

**Comment by CrabDude** [comment]
Currently, `babel-node-debug` adds a breakpoint via V8 Debugger Protocol pattern matching on the main file's name, then clears it once it's hit. It's a total hack, but based on the current architecture of node.js there's not a better way to do this without updating core.

**Comment by CrabDude** [comment]
There's also the additional caveat that I've not yet figured out how to add the breakpoint to the uncompiled (babel/es6/es7) code. I've raised on issue on https://github.com/node-inspector/node-inspector/issues/748 in search of recommendations to accomplish this.

**Comment by sebmck** [comment]
I'm not sure what you're requesting. What exactly do you want to be changed?

**Comment by CrabDude** [comment]
I'll update the OP once I've explained successfully.

**WHAT** is the problem?

**Steps to reproduce:**

1. Start script with `babel-node` and `--debug-brk`:

    ```bash
    $ babel-node --debug-brk foo.js`
    ```
1. Start any V8 debugger (`node-inspector` is the most common): 

    ```bash
    $ node-inspector`
    ```

**Problem:** The first breakpoint when the debugger starts is on line 1 of `babel/lib/_babel-node.js`, which I believe is this file: [packages/babel-cli/src/_babel-node.js](https://github.com/babel/babel/blob/master/packages/babel-cli/src/_babel-node.js).

**Expected:** The same behavior as `node --debug-brk foo.js`: the first breakpoint when the debugger starts is on line 1 of foo.js.

---

**WHY does this need to be fixed in babel?**

I solved a similar problem when creating `babel-node-debug` (a `babel-node` powered equivalent to `node-inspector`'s `node-debug` CLI), but it's impossible for technical / logistical reasons to solve the above `babel-node --debug-brk foo.js` scenario outside of babel itself.

**Which is what bring me here.**

There's really no reason this shouldn't be expected behavior in `babel-node`, especially since debugging is currently a major pain point for babel development (I'm actually presenting [a 3 hour session at the upcoming "Node Community Convention" conference in SF on this](http://nodecommunityconvention.com/san-francisco-2016/agenda/day-one-adenda/900am-b.html)). So if babel addressed this, it would solve the V8 debugger initial breakpoint issue (not specific to `node-inspector`) and obviate the need for `babel-node-debug` by allowing:

```bash
$ node-debug --nodejs `which babel-node` foo.js
```

---

**HOW would babel solve this?**

There're 2 solutions I can think of  when `--debug-brk` is passed:

1. **Simplest & best:** Prepend the main file with `debugger;\n`
1. Connect to the V8 Debug Server process to programmatically clear and set the appropriate breakpoints.

**The first is almost trivial, and I'm happy to submit a PR if it would be accepted.**

The second is more complex, but doable since you have the path to the main file and it's what I'm doing in `babel-node-debug` with the caveat that there doesn't seem to be a way in the V8 Debug Protocol to add a breakpoint to the uncompiled (sourcemap) version of the file programatically.

**Comment by gear54rus** [comment]
+1 on this.

It seems logical that `--debug-brk` should be handled separately (instead of passing it to `node` directly).

The workflow described in OP will help debug things like `child_process`es running `babel-node`.

**Comment by CrabDude** [comment]
I was thinking about this a bit more, and I think the ideal solution may be a mix of the two proposed solutions.

Some thoughts:

1. The "prepend" solution is much slower to startup (~60s) because certain things are delayed until a debugger connects.
1. We can eliminate this startup delay almost entirely by connecting our own right away and issuing a "continue" command to the V8 Debugger Server.

So, if we combine the prepend `debugger;` solution with the connect to the V8 Debugger Server and issue "continue" to skip the `--debug-brk` "breakpoint" then we get the best of both worlds:

**Fast debugger startup on L1 of our main uncompiled file.**

Also, for anyone who's not familiar, `babel-node` cannot just pass `--debug` to node instead of `--debug-brk` for the prepend solution because breakpoints are ignored in `--debug` mode if a debugger is not already connected, so by the time any debugger connects (apart from the debugger that babel connects to to issue the "continue" command), it's usually too late and the application has run to completion.

**Comment by CrabDude** [comment]
FWIW, there's a much better way to debug babel now:

```
node -r babel-register --inspect --debug-brk foo.js
```

Which is transparent with this alias:

```
alias bode='node -r babel-register'
```

Or a bash passthru like:

```bash
exec /usr/bin/env node -r babel-register "$@"
```

Which I packaged on npm for the world to enjoy: [`bode`](http://npmjs.com/package/bode).

**Comment by loganfsmyth** [comment]
What you described seems to be another approach to the `babel-node` command that is included in `babel-cli`.

**Comment by CrabDude** [comment]
@loganfsmyth Except it solves this issue, which is specifically [about how `babel-node` does not](https://github.com/babel/babel/issues/2565#issue-112484238).

**Comment by CrabDude** [comment]
Also, note that apart from solving the OP (first breakpoint line being application code, not babel-node code):

1. It's far simpler and easier than the hoops I have to jump through with [`babel-node-debug` (read the code; it's a hack)](https://github.com/CrabDude/babel-node-debug/blob/master/bin/bode-debug.js).
1. `exec` is transparent, whereas `babel-node` intercepts and consume several arguments, requiring the non-transparent use of an extra ` -- `
1. It's significantly faster, b/c it doesn't require a parent (second) process.

Granted, it's a shell script, so windows is out, but you can't win 'em all. I suppose I could add a build step in `bode` that toggles to windows compatible version if someone wanted to write it.
